### PR TITLE
storage/static.go: storage backend should not explicitly lower-case email ids.

### PR DESCRIPTION
**Pull request by rithujohn191**

### CODE REVIEW

File: storage/static.go [hunk 1]
```diff
@@ -91,8 +91,9 @@ type staticPasswordsStorage struct {
 func WithStaticPasswords(s Storage, staticPasswords []Password) Storage {
 	passwordsByEmail := make(map[string]Password, len(staticPasswords))
 	for _, p := range staticPasswords {
-		p.Email = strings.ToLower(p.Email)
-		passwordsByEmail[p.Email] = p
+		//Enable case insensitive email comparison.
+		lowerEmail = strings.ToLower(p.Email)
```
Comment (ericchiang) [inline storage/static.go]: :=

**Comment by rithujohn191** [comment]
added a test and documentation

File: storage/memory/static_test.go [hunk 2]
```diff
@@ -173,10 +174,14 @@ func TestStaticPasswords(t *testing.T) {
...
 		},
 		{
-			name: "create password",
+			name: "create passwords",
 			action: func() error {
-				return s.CreatePassword(p3)
+				if err := s.CreatePassword(p3); err != nil {
+					return err
+				}
+				return s.CreatePassword(p4)
```
Comment (ericchiang) [inline storage/memory/static_test.go]: Need to read this value back an ensure the emails match.

File: storage/static.go [hunk 3]
```diff
@@ -91,8 +92,9 @@ type staticPasswordsStorage struct {
 func WithStaticPasswords(s Storage, staticPasswords []Password) Storage {
 	passwordsByEmail := make(map[string]Password, len(staticPasswords))
 	for _, p := range staticPasswords {
-		p.Email = strings.ToLower(p.Email)
-		passwordsByEmail[p.Email] = p
+		//Enable case insensitive email comparison.
+		lowerEmail := strings.ToLower(p.Email)
+		passwordsByEmail[lowerEmail] = p
```
Comment (kans) [inline storage/static.go]: Could we log or something if the `lowered` email as already in the map?

Comment (ericchiang) [inline storage/static.go]: We would just return an error here. @rithujohn191 what do you think?

File: storage/memory/static_test.go (hunk 2 above)
Comment (rithujohn191) [inline storage/memory/static_test.go]: This test makes sure that "Spam@example.com" cannot be created, so it makes sure an error is returned

Comment (rithujohn191) [inline storage/memory/static_test.go]: do you want me to check if "spam@example.com" gets created with the all lower case?

Comment (ericchiang) [inline storage/memory/static_test.go]: Something like

```
p, err := s.GetPassword(strings.ToLower(p4.Email))

p.Email == p4.Email
```

File: storage/static.go (hunk 3 above)
Comment (ericchiang) [inline storage/static.go]: Actually I only think we can log this error. Erroring out would be backward incompatible, if sensible :(

Comment (rithujohn191) [inline storage/static.go]: So if we don't error out we essentially allow the user to overwrite the map. For the following ConfigMap:
```
staticPasswords:
- email: "admin@example.com"
  # bcrypt hash of the string "password"
  hash: "$2a$10$2b2cU8CPhOTaGrs1HRQuAueS7JTT5ZHsHSzYiFPm1leZck7Mc8T4W"
  username: "admin"
  userID: "08a8684b-db88-4b73-90a9-3cd1661f5466"
- email: "Admin@example.com"
  # bcrypt hash of the string "password"
  hash: "$2a$10$2b2cU8CPhOTaGrs1HRQuAueS7JTT5ZHsHSzYiFPm1leZck7Mc8T4W"
  username: "Admin"
  userID: "08a8684b-db88-4b73-90a9-3cd1661f5466"
```

Only the second password gets registered. And then if the user tries to login with the first email it still succeed but returns the wrong email id in the claims:

```
"email": "Admin@example.com",
```

That's not desirable behavior :(

Comment (ericchiang) [inline storage/static.go]: @rithujohn191 right, but prior to this that config would not have errored, if we error now we might break someone during a dex upgrade.

Comment (rithujohn191) [inline storage/static.go]: logged the error

File: storage/memory/static_test.go (hunk 2 above)
Comment (rithujohn191) [inline storage/memory/static_test.go]: done

**Comment by rithujohn191** [comment]
working on adding the logger

**Comment by rithujohn191** [comment]
Done :)

**Pull request merged**